### PR TITLE
Feature: Get property selection mapping

### PR DESCRIPTION
**Pull request by Jan-Panasiam**
Pull the property selection data from the Plentymarkets Rest API and
create a dictionary with a property ID -> selection ID -> language ->
value structure to enable the user to get the appropriate values for
their selection IDs when pulling their variation property data.

Signed-off-by: Jan <jan.sallermann@panasiam.de>

### CODE REVIEW

File: plenty_api/api.py [hunk 1]
```diff
@@ -837,6 +839,48 @@ def plenty_api_get_contacts(self,
             refine=refine,
             additional=additional)
 
+    def plenty_api_get_property_selections(self,
+                                           refine: dict = None,
+                                           additional: list = None):
```
Comment (initBasti) [inline plenty_api/api.py:844]: You define additional but you have not defined any valid additional values.

File: plenty_api/api.py [hunk 2]
```diff
@@ -837,6 +839,48 @@ def plenty_api_get_contacts(self,
...
+        """
+        Get maaping of selection IDs to the actual value in all avaiable
```
Comment (initBasti) [inline plenty_api/api.py:846]: s/maaping/mapping/

File: plenty_api/api.py [hunk 3]
```diff
@@ -837,6 +839,48 @@ def plenty_api_get_contacts(self,
...
+        languages.
+
+        Parameter:
+            refine      [dict]  -   Apply filters to the request
+                                    Example:
+                                    {'email': 'a@posteo.net', 'name': 'Thomas'}
```
Comment (initBasti) [inline plenty_api/api.py:852]: Strange example those are not actual examples for the refine attributes of this call. You pretty much only explain how a dictionary looks, which is not very helpful.

File: plenty_api/api.py [hunk 4]
```diff
@@ -837,6 +839,48 @@ def plenty_api_get_contacts(self,
...
+            additional  [list]  -   Add additional elements to the response
+                                    data.
+                                    Example:
+                                    ['addresses']
```
Comment (initBasti) [inline plenty_api/api.py:856]: Same here please do not give examples that are irrelevant to the call.

File: plenty_api/api.py [hunk 5]
```diff
@@ -837,6 +839,48 @@ def plenty_api_get_contacts(self,
...
+        )
+
+        selection_map = {}
+        for selection in data:
+            key = selection['propertyId']
+            value_id = selection['id']
+            if not key in selection_map.keys():
+                selection_map[key] = {}
+
+            selection_map[key][value_id] = {}
```
Comment (initBasti) [inline plenty_api/api.py:873]: This is good use-case for a defaultdict:
```
selection_map = defaultdict(lambda: defaultdict(dict))
```

File: plenty_api/api.py [hunk 6]
```diff
@@ -837,6 +839,48 @@ def plenty_api_get_contacts(self,
...
+                        [dict]  -   Mapping of all selection property values
+        """
+        if self.data_format == 'dataframe':
+            logging.error("Mapping can only be created with the JSON format")
+            return
+
+        data = self.__plenty_api_generic_get(
+            domain='property',
+            path='/selections',
+            refine=refine
```
Comment (initBasti) [inline plenty_api/api.py:863]: The `/rest/properties/selections` route does not accept parameters, the way to refine the result is by using the `/rest/properties/selections` when no refine argument is defined and `​/rest​/v2​/properties​/{propertyId}​/selections` when one property ID is given. You can also think about accepting a list of property IDs and then say len(list) < 4 => loop through the single requests else request all and filter the IDs out.

**Review by initBasti — CHANGES_REQUESTED** [review]
Overall looks quite good, but please fix the mentioned problems.

File: plenty_api/api.py (hunk 6 above)
Comment (Jan-Panasiam) [reply plenty_api/api.py:863]: It actually does, at least in this implementation if you give the refine argument into the function {'propertyId': 121} it only requests and also only give you the values of the propertyId 121, must not be properly documented in the PlentyMarkets documentation, or it is what the call PropertySelectionId, but it still works.

File: docs/API_reference.md [hunk 7]
```diff
@@ -275,6 +276,23 @@ There are currently two supported output formats: 'json' and 'dataframe'.
 The 'json' format simply returns the raw response, without page information and with multiple pages combined into a single data structure.  
 The 'dataframe' format transforms that data structure into a pandas DataFrame, which contains subparts in json, that can be split further by the user application.
 
+---
+
+##### plenty_api_get_property_selections: <a name='get-property-selections'></a>
+
+Return a mapping of Selection values for selction properties.
```
Comment (initBasti) [inline docs/API_reference.md:283]: s/Selection values/seleciton values/
s/selction/selection/

But actually, I would reword this to:
`Create a mapping of property IDs to the selection-values of the property with the ID of the selection and text values for all available languages.`

File: docs/API_reference.md [hunk 8]
```diff
@@ -275,6 +276,23 @@ There are currently two supported output formats: 'json' and 'dataframe'.
...
+
+[*Optional parameter*]:
+
+The **refine** field can be used to reduce the request to specific property IDs.
+example: refine={'propertyID': 123}
```
Comment (initBasti) [inline docs/API_reference.md:288]: So the syntax for multiple IDs is: `{'propertyID': [123, 234]}` ? Or is this only possible for one value? If the latter is the case then please reword the sentence above accordingly. If the former is correct, then please add an additional example highlighting how to work with multiple IDs.

File: docs/API_reference.md [hunk 9]
```diff
@@ -275,6 +276,23 @@ There are currently two supported output formats: 'json' and 'dataframe'.
...
+
+[*Output format*]:
+
+There are currently two supported output formats: 'json' and 'dataframe'.
```
Comment (initBasti) [inline docs/API_reference.md:292]: You actually only support json, currently.
You could implement dataframe from what you got like this:
dict version:
`{1: {11: {'de': 'value_1', 'en': 'value_2'}}, 2: {12: {'de': 'value_3', 'en': 'value_4'}}}`
dataframe version:
`DataFrame([[1, 11, 'de', 'value_1'], [1, 11, 'en', 'value_2'], [2, 12, 'de', 'value_3'], [2, 12, 'en', 'value_4']], columns=['property_id', 'selection_id', 'lang', 'value'])`

File: docs/API_reference.md [hunk 10]
```diff
@@ -275,6 +276,23 @@ There are currently two supported output formats: 'json' and 'dataframe'.
...
+
+[*Optional parameter*]:
+
+The **refine** field can be used to reduce the request to specific property IDs.
+example: refine={'propertyID': 123}
+
+[*Output format*]:
+
+There are currently two supported output formats: 'json' and 'dataframe'.  
+The 'json' format returns a dictionary containing the property_id, selection_id, language and finally the value.
```
Comment (initBasti) [inline docs/API_reference.md:293]: I find this description of the dictionary a bit vague because this sounds like:
{'property_id': 123, 'selection_id': 12, 'language': 'de', 'value': 'value_1'}
but it is:
`{123: {12: {'de': 'value_1'}}}`
So maybe a better description is:
... returns a dictionary of property IDs to its available selection values

And maybe add a small sample result, so that the user knows what to expect.

File: docs/API_reference.md [hunk 11]
```diff
@@ -275,6 +276,23 @@ There are currently two supported output formats: 'json' and 'dataframe'.
...
+The 'dataframe' format transforms that data structure into a pandas DataFrame, which contains subparts in json, that can be split further by the user application.
```
Comment (initBasti) [inline docs/API_reference.md:294]: As described above I believe we can present the user with a better dataframe.

File: plenty_api/api.py [hunk 12]
```diff
@@ -837,6 +840,36 @@ def plenty_api_get_contacts(self,
             refine=refine,
             additional=additional)
 
+    def plenty_api_get_property_selections(self, refine: dict = None):
+        """
+        Get mapping of selection IDs to the actual value in all avaiable
```
Comment (initBasti) [inline plenty_api/api.py:845]: s/Get mapping/Get a mapping/
s/avaiable/available/

File: plenty_api/api.py (hunk 6 above)
Comment (initBasti) [reply plenty_api/api.py:858]: Alright, we should highlight that we diverge from the official documentation in this case within our documentation.

**Review by initBasti — CHANGES_REQUESTED** [review]

**Review by initBasti — APPROVED** [review]
Looks good thank you for the changes.
